File: 2018/src/day03.rs
use std::collections::HashMap;
use std::fs;

pub fn day03() {
    let input = "data/input-03.txt";
    let file = fs::read_to_string(input)
        .unwrap()
        .replace('x', ",")
        .replace(": ", ",");
    let mut vals = Vec::new();
    for line in file.lines() {
        let line: Vec<_> = line
            .split("@ ")
            .last()
            .unwrap()
            .split(',')
            .map(|z| z.parse::<i32>().unwrap())
            .collect();
        vals.push(line);
    }

    let mut patterns = HashMap::new();
    for val in &vals {
        if let [lef, top, wid, hei] = val[..] {
            for i in lef..(lef + wid) {
                for j in top..(top + hei) {
                    *patterns.entry((i, j)).or_insert(0) += 1;
                }
            }
        }
    }
    let res = patterns.values().fold(0, |acc, &z| acc + (z > 1) as i32);
    println!("Part 1: {res}");
    assert_eq!(res, 111326);

    let mut res = 0;
    'outer: for (idx, val) in vals.iter().enumerate() {
        if let [lef, top, wid, hei] = val[..] {
            for i in lef..(lef + wid) {
                for j in top..(top + hei) {
                    if patterns.get(&(i, j)) > Some(&1) {
                        continue 'outer;
                    }
                }
            }
        }
        res = idx + 1;
        break;
    }
    println!("Part 2: {res}");
    assert_eq!(res, 1019);
}

File: 2018/nim/day03.nim
import sequtils, sets, strutils, sugar, tables

let vals = collect(newSeq):
    for line in lines("data/input-03.txt"):
        var split = line.split(" @ ")
        let issue = split[0].replace("#", "").parseInt()
        split = split[1].split(": ")
        let coor = split[0].split(",").map(parseInt)
        let size = split[1].split("x").map(parseInt)
        (issue, coor, size)

var cover: CountTable[(int, int)]
var claim: Table[(int, int), seq[int]]
for (issue, coor, size) in vals:
    let (lef, top) = (@coor[0], @coor[1])
    let (wid, hei) = (@size[0], @size[1])
    for x in lef..<lef + wid:
        for y in top ..< top + hei:
            let val = (x, y)
            if cover.contains(val):
                cover.inc(val)
                claim[val].add(issue)
            else:
                cover.inc(val)
                claim[val] = @[issue]

var res1 = 0
for i in cover.values():
    res1 += int(i > 1)
echo "Part 1: ", res1
assert(res1 == 111326)

let used = collect(initHashSet()):
    for (k, v) in cover.pairs():
        if v > 1:
            for c in claim[k]: {c}
let issues = collect(initHashSet()):
    for v in vals:
        {v[0]}
let res2 = issues.difference(used).toSeq()[0]
echo "Part 2: ", res2
assert(res2 == 1019)
